LoopVectorize: fix phi cost when it is scalar after vectorization

artagnon opened the pull request:
In getInstructionCost, when a phi node is identified as a fixed-order recurrence, and the vectorization factor indicates that we're vectorizing, we return the cost of a shuffle. However, if it also satisfies the condition isScalarAfterVectorization, attempting to get the shuffle cost will result in a crash from a cast. Fix this.

Fixes #72969.

-- 8< --
Based on #74111.

david-arm reviewed (approved):
LGTM! Thanks for the fix.

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/pr72969.ll:
```diff
@@ -0,0 +1,99 @@
+; CHECK-NEXT:    [[P1:%.*]] = ptrtoint ptr [[P]] to i64
+; CHECK-NEXT:    [[TMP0:%.*]] = add i64 [[P1]], 8
+; CHECK-NEXT:    [[UMAX2:%.*]] = call i64 @llvm.umax.i64(i64 [[TMP0]], i64 add (i64 ptrtoint (ptr @h to i64), i64 1))
+; CHECK-NEXT:    [[TMP1:%.*]] = add i64 [[UMAX2]], -5
```
Is it possible to adjust the test so that no SCEV checks are neede?

artagnon replied:
Actually, since there is a minimal test case for the crash in X86, I think we can remove this file altogether. Note that the X86 test [doesn't vectorize anything](https://github.com/llvm/llvm-project/pull/74456/files#diff-8b1725ba52f1290cb42ea43d60ea58d638a3b14ae3063e6c7e8caa9fda8a3438), since I've disabled the vector extensions in the RUN line.

fhahn commented:
> However, if it also satisfies the condition isScalarAfterVectorization, this cost is incorrect, as it won't be converted into a shuffle. Fix the cost.

Hmm, is this correct? It looks like in the VPlan, there's 

```
FIRST-ORDER-RECURRENCE-PHI ir<%idx.ext.merge> = phi ir<1>, ir<%idx>
EMIT vp<%8> = first-order splice ir<%idx.ext.merge>, ir<%idx>
```

so a vector phi and shuffle should be generated. It may be good to also add a use of the recurrence to the test, the load and GEP will get DCE'd.

artagnon committed:
LoopVectorize: add test for crash in #72969

artagnon commented:
Fixed commit message.

> It may be good to also add a use of the recurrence to the test, the load and GEP will get DCE'd.

I don't think this would apply, since nothing gets vectorized in the minimal X86 example, and I don't think we need an example showing vectorization to fix the bug.

artagnon committed:
LoopVectorize: fix phi cost when it is scalar after vectorization

In getInstructionCost, when a phi node is identified as a fixed-order
recurrence, and the vectorization factor indicates that we're
vectorizing, we return the cost of a shuffle. However, if it also
satisfies the condition isScalarAfterVectorization, attempting to get
the shuffle cost will result in a crash from a cast. Fix this.

Fixes #72969.

artagnon commented:
@fhahn Ping. Are you happy with the changes?

fhahn commented:
> @fhahn Ping. Are you happy with the changes?

What are your thoughts about the below? For that, we would need a case that gets vectorized with the change.

>> Hmm, is this correct? It looks like in the VPlan, there's
>> 
>>      FIRST-ORDER-RECURRENCE-PHI ir<%idx.ext.merge> = phi ir<1>, ir<%idx>
>>      EMIT vp<%8> = first-order splice ir<%idx.ext.merge>, ir<%idx>
>>
>> so a vector phi and shuffle should be generated. 
>> It may be good to also add a use of the recurrence to the test, the load and GEP will get DCE'd.
> I don't think this would apply, since nothing gets vectorized in the minimal X86 example, and I don't think we need an example showing vectorization to fix the bug.

Ideally we would have a test for both cases that are impacted by the patch: vectorizing with the new cost and also not vectorizing (and not crashing).

artagnon commented:
> Ideally we would have a test for both cases that are impacted by the patch: vectorizing with the new cost and also not vectorizing (and not crashing).

Okay, I struggled a lot with your requests, because the bug is very subtle:
1. It's impossible to get a test under X86 to first crash, and then be vectorized after the fix, since vectorization depends on sse4.2, and the bug isn't triggered when sse4.2 is enabled.
2. I tried everything, but I think it's impossible to remove the SCEV checks.
3. I tried to use the result of the recurrence in a lot of ways, but none of them get the test to crash first, and then be vectorized after the fix.

I think we have no choice but to revert to my original version of X86 and AArch64 tests.

artagnon commented:
@fhahn This is the best I could do (see previous comment). Do you think it's okay to merge?

artagnon commented:
Thanks @fhahn: the patch is now ready. Your comment about replacing the `load` with the `store` makes sense, but can you clarify why you thought `-force-vector-width=4` would work? Also, if you know, what is the difference between sse and sse4.2 in the context of this test?

fhahn reviewed (commented):
> Thanks @fhahn: the patch is now ready. Your comment about replacing the `load` with the `store` makes sense, but can you clarify why you thought `-force-vector-width=4` would work? Also, if you know, what is the difference between sse and sse4.2 in the context of this test?

> Thanks @fhahn: the patch is now ready. Your comment about replacing the load with the store makes sense, but can you clarify why you thought -force-vector-width=4 would work? Also, if you know, what is the difference between sse and sse4.2 in the context of this test?

The reason for not vectorizing was probably the cost-model deciding it is not profitable; `-force-vector-width=4` side-steps that by forcing to vectorize with VF=4 if legally possible.

Not sure about `sse4.2`.

artagnon commented:
Will author a fresh patch soon.
